### PR TITLE
improve substitutions (#557)

### PR DESCRIPTION
**Pull request by harakas**

### CODE REVIEW

File: symengine/tests/basic/test_subs.cpp [hunk 1]
```diff
@@ -103,6 +103,20 @@ TEST_CASE("Add: subs", "[subs]")
     r1 = add(mul(x, add(integer(3), sin(integer(4)))), sin(integer(1)));
     r2 = add(add(integer(3), sin(integer(4))), sin(integer(1)));
     REQUIRE(eq(*r1->subs(d), *r2));
+
+    // (2+2*x).subs({2: y}) -> y+2*x
```
Comment (certik) [inline symengine/tests/basic/test_subs.cpp]: Shouldn't this return `y+y*x`?

Comment (harakas) [inline symengine/tests/basic/test_subs.cpp]: Well spotted. I neglected to substitute the term coefficients.

File: symengine/add.cpp [hunk 2]
```diff
@@ -337,26 +337,42 @@ RCP<const Basic> Add::subs(const map_basic_basic &subs_dict) const
...
-            Add::dict_add_term(d, p.second, p.first);
-        } else if (is_a<Integer>(*term) and
-                rcp_static_cast<const Integer>(term)->is_zero()) {
-            continue;
-        } else if (is_a_Number(*term)) {
-            iaddnum(outArg(coef),
-                    mulnum(p.second, rcp_static_cast<const Number>(term)));
-        } else if (is_a<Add>(*term) and p.second->is_one()) {
+
+    auto add_cterm = [&](const RCP<const Number>& c, const RCP<const Basic>& term) {
```
Comment (isuruf) [inline symengine/add.cpp]: Can you move this code to a new static method like 
```
static void Add::dict_add_term_new(const Ptr<RCP<const Number>> &coef, umap_basic_num &d, 
        const RCP<const Number> &coef, const RCP<const Basic> &t);
```
This method is duplicated also in `Add::diff` and we can use it there as well.

**Comment by isuruf** [comment]
+1 to merge

**Comment by harakas** [comment]
Thinking about this one finds problems. Another problem case: `(-2-2*x).subs({2:y})` does no substitutions. But how far do you want to take it, say in sympy you can do `(2+2*x).subs(-2,y) -> -x*y-y`. And what about minus one inlining. With `(x-y).subs(y,z+2)` -- is that `x-(z+2)`  (current symengine, no inlining) or `x-z-2` (in sympy)?

**Comment by certik** [comment]
Well, in order to answer how far we want to take it, we should have an equivalent of something like `ReplaceAll` in Mathematica:
```
https://reference.wolfram.com/language/ref/ReplaceAll.html
```
I assume the SymPy version is pretty close.

In addition however, we also want to have a much simpler and faster version, which will not be as thorough, but it would run fast, and you can use it when the simpler method is enough and speed matters.

**Comment by harakas** [comment]
I don't have access to Mathematica or have any experience with it. I assume in case of subs() here we are talking about the simple fast version and should be looking for practical ideas and limits.

Sign juggling at its most primitive implementation would slow things up by double, but it can be improved by say creating a new dictionary that contains also the negative substitutions within it. Is that acceptable for our simple and fast version or is it already beyond it? Now that I've thought about it, I think that it should not be part of subs(), but an external measure.

So for now I believe this change is quite sufficient.

**Comment by certik** [comment]
I agree. I think it looks good, so we can merge it.

**Pull request merged**